questions/250/explanation.md

```diff
@@ -1,62 +1,61 @@
-
-We now do deduction of  `foo₂(T...)` from `foo₁(X)`, which deduces `T=X`. We then do deduction of `foo₁(T)` from `foo₂(Y)`. It looks like we would get `T=Y`, but §[temp.deduct.type]¶10 says:
-
-> During partial ordering, if `Ai` was originally a function parameter pack [and] `Pi` is not a function parameter pack, template argument deduction fails.
-
-`Ai` in this case is `Y`, which was originally a function parameter pack. `Pi` is `T`, which is not a function parameter pack, so deduction fails.
-
-Since we can deduce `foo₂` from `foo₁`, but not `foo₁` from `foo₂`,  `foo₁` is more specialized and a better function. Overload resolution picks it and prints `A`.
-
-### `foo(1,2)`:
-
-Now let's look at the call `foo(1,2)`.
-
-Again we start by deducing the template arguments. For `foo₁`, the first argument `1` is used to deduce `T` to be `int`. It's signature is then `foo₁(int, ...)`.
-
-For `foo₂`, `T...` is a function parameter pack. The arguments `1` and `2` are used to deduce this as two `int`s, and it's signature is `foo₂(int, int)`.
-
-Which of these is the best match? We turn again to §[over.match.best]¶1 which checks whether
-
->for some argument *j*, `ICSj(F1)` is a better conversion sequence than `ICSj(F2)`
-
-For both templates, the first argument is an exact match. For `foo₁`, the second argument `2` requires an ellipsis conversion sequence to match the second parameter `...`: §[over.ics.ellipsis]¶1:
-
-> An ellipsis conversion sequence occurs when an argument in a function call is matched with the ellipsis parameter specification of the function called
-
+The first `foo` (let's call it `foo₁`) is a template for an old-school (pre variadic templates) variadic function. It takes a first argument of type `T`, followed by a varying number of arguments.
+
+The second `foo` (let's call it `foo₂`) is a variadic function template (notice the ellipsis in `typename...`). It takes a parameter pack.
+
+Which overload is selected for each of the calls to `foo`?
+
+When a function is overloaded, we first need to find which of the functions are viable for a given function call. Then, we need to find out which of those viable functions is the best one. For function templates, we first perform template argument deduction to generate candidate specializations.
+
+### `foo(1)`:
+
+First let's look at the call `foo(1)`. The explanation is lengthy, but rest assured we reuse most of it in the much shorter explanation of `foo(1,2)`! 
+
+We start by deducing the template arguments for both overloads. For `foo₁`, the argument `1` is used to deduce `T` to be `int`. It's signature is then `foo₁(int, ...)`. However, for overload resolution we ignore the `...` since it doesn't have a matching argument: §[over.match.viable]¶2.3:
+
+> For the purposes of overload resolution, the parameter list is truncated on the right, so that there are exactly *m* parameters.
+
+So we end up with the viable function `foo₁(int)`.
```

Comment:
```suggestion
We start by deducing the template arguments for both overloads. For `foo₁`, the argument `1` is used to deduce `T` to be `int`. Its signature is then `foo₁(int, ...)`. Is it viable? §[over.match.viable]¶2.1:

> — If there are `m` arguments in the list, all candidate functions having exactly `m` parameters are viable.

But don't we have 2 parameters and 1 argument? It turns out an ellipsis is not considered to be a parameter in this context since it's not included in the *parameter-declaration* grammar production (§[dcl.fct]¶nt:parameter-declaration).

Regarding §[over.match.viable]¶3 and §[over.match.viable]¶4, `foo₁` doesn't have any constraints and there exists an implicit conversion sequence from `1` to `int` (namely, an empty standard conversion sequence since no conversions are required). So `foo₁` is viable.
```